implementations/rust/ockam/ockam_command/src/util/installer.rs

```diff
@@ -0,0 +1,181 @@
+use crate::error::Result;
+use clap::crate_version;
+use miette::miette;
+use miette::Context;
+use miette::IntoDiagnostic;
+use std::process::Stdio;
+use std::{path::PathBuf, process::Command};
+
+fn upgrade_with_brew() -> miette::Result<()> {
+    let result = Command::new("brew")
+        .args(["upgrade", "ockam"])
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .output()
+        .into_diagnostic()
+        .context("Failed to upgrade ockam with homebrew")?;
+    if result.status.success() {
+        Ok(())
+    } else {
+        Err(miette!("Failed to upgrade ockam here"))
+    }
+}
+
+fn check_installed_with_brew(current_version: &str) -> bool {
+    // if brew finds ockam and it is the same version as the current version, then
+    // ockam has been installed with brew and so we upgrade with brew
+    let brew_result = Command::new("brew")
+        .args(["ls", "--versions", "ockam"])
+        .output();
+    if let Ok(output) = brew_result {
+        if !output.stdout.is_empty() {
+            let brew_version_result = String::from_utf8(output.stdout);
+            if let Ok(brew_version) = brew_version_result {
+                if brew_version.trim().contains(current_version) {
+                    return true;
+                }
+            }
+        }
+    }
+    false
+}
+
+fn uninstall_with_brew() -> miette::Result<()> {
+    let result = Command::new("brew")
+        .args(["uninstall", "ockam"])
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .output()
+        .into_diagnostic()
+        .context("Failed to uninstall ockam with homebrew")?;
+    if result.status.success() {
+        Ok(())
+    } else {
+        Err(miette!("Failed to uninstall ockam with homebrew"))
+    }
+}
+
+fn get_ockam_home() -> Result<PathBuf> {
+    // check to see if OCKAM_HOME is set
+    let ockam_home = std::env::var("OCKAM_HOME");
+    if let Ok(home) = ockam_home {
+        let mut pathbuf = PathBuf::new();
+        pathbuf.push(home);
+        return Ok(pathbuf);
+    }
+
+    // check to see if it is using the default ockam home
+    let home = std::env::var("HOME");
+    if let Ok(home) = home {
+        let mut pathbuf = PathBuf::new();
+        pathbuf.push(home);
+        pathbuf.push(".ockam");
+        if pathbuf.exists() && pathbuf.is_dir() {
+            return Ok(pathbuf);
+        }
+    }
+    Err(miette!("Failed to get ockam home").into())
+}
+
+fn delete_binary() -> miette::Result<()> {
+    let binary_path = std::env::current_exe().into_diagnostic()?;
+    std::fs::remove_file(binary_path)
+        .into_diagnostic()
+        .context("Failed to delete binary")
+}
+
+fn upgrade_binary(version: &str) -> miette::Result<()> {
+    let ockam_home = get_ockam_home()?;
+    let ockam_home = ockam_home.to_str().expect("Invalid ockam home");
+    delete_binary()?;
```

Comment:
Perhaps this should rename the old binary until the install script has successfully run, and then delete it? That way a failed upgrade doesn't leave the user with no ockam binary, which it could with this approach.

Reply:
Yeah I was thinking about solving it like that as well, I'll make those changes.